source/hp/dof_handler.cc

```diff
@@ -1745,55 +1724,169 @@ namespace hp
-                for (unsigned int i = 0; i < cell->n_children(); ++i)
-                  if (cell->child(i)->is_locally_owned())
-                    cell->child(i)->set_active_fe_index(
-                      levels[cell->level()]->active_fe_index(cell->index()));
+                // Store the active_fe_index of each cell that will be refined
+                // to and distribute it later on its children.
+                refined_cells_fe_index.push_back(
+                  {cell->id(), cell->active_fe_index()});
+              }
+            else if (cell->coarsen_flag_set())
+              {
+                // From all cells that will be coarsened, determine their parent
+                // and calculate its proper active_fe_index, so that it can be
+                // set after refinement.
+                const auto &parent = cell->parent();
+                // Check if the active_fe_index for the current cell has been
+                // determined already.
+                if (std::find_if(
+                      coarsened_cells_fe_index.cbegin(),
+                      coarsened_cells_fe_index.cend(),
+                      [&parent](
+                        const std::pair<CellId, unsigned int> &element) {
+                        return element.first == parent->id();
+                      }) == coarsened_cells_fe_index.cend())
+                  {
+                    std::set<unsigned int> fe_indices_children;
+                    for (unsigned int child_index = 0;
+                         child_index < GeometryInfo<dim>::max_children_per_cell;
+                         ++child_index)
+                      fe_indices_children.insert(
+                        parent->child(child_index)->active_fe_index());
+
+                    const unsigned int fe_index =
+                      fe_collection.find_least_dominating_fe_in_collection(
+                        fe_indices_children, /*codim=*/0);
+
+                    Assert(
+                      fe_index != numbers::invalid_unsigned_int,
+                      ExcMessage(
+                        "No FiniteElement has been found in your FECollection "
+                        "that dominates all children of a cell you are trying "
+                        "to coarsen!"));
+
+                    coarsened_cells_fe_index.push_back(
+                      {parent->id(), fe_index});
+                  }
               }
           }
+
+        // Finally, restore current subdomain_ids.
+        if (shared_tria != nullptr)
+          if (shared_tria->with_artificial_cells())
+            {
+              typename parallel::shared::Triangulation<dim, spacedim>::
+                active_cell_iterator cell = shared_tria->begin_active(),
+                                     endc = shared_tria->end();
+
+              for (unsigned int index = 0; cell != endc; ++cell, ++index)
+                cell->set_subdomain_id(saved_subdomain_ids[index]);
+            }
```

Comment:
I think it would be useful to delete the `active_fe_index` again here from artificial cells. That's because these have a tendency to become outdated and so removing them avoids people accidentally relying on them.